libs/payments/paypal/src/lib/paypal.service.spec.ts

```diff
@@ -0,0 +1,126 @@
+import { Test } from '@nestjs/testing';
+import { faker } from '@faker-js/faker';
+
+import { CurrencyManager } from '@fxa/payments/currency';
+import {
+  AccountDbProvider,
+  MockAccountDatabaseNestFactory,
+} from '@fxa/shared/db/mysql/account';
+
+import { PayPalManager } from './paypal.manager';
+import { PayPalService } from './paypal.service';
+import { BillingAgreement, BillingAgreementStatus } from './paypal.types';
+import { PaypalCustomerManager } from './paypalCustomer/paypalCustomer.manager';
+
+describe('PayPalService', () => {
+  let paypalService: PayPalService;
+  let paypalManager: PayPalManager;
+  let currencyManager: CurrencyManager;
+  let paypalCustomerManager: PaypalCustomerManager;
+
+  beforeAll(async () => {
+    const moduleRef = await Test.createTestingModule({
+      providers: [
+        PayPalService,
+        PayPalManager,
+        CurrencyManager,
+        PaypalCustomerManager,
+        {
+          provide: AccountDbProvider,
+          useValue: MockAccountDatabaseNestFactory,
+        },
+      ],
+    })
+      .overrideProvider(PayPalManager)
+      .useValue({
+        createBillingAgreement: jest.fn(),
+        getBillingAgreement: jest.fn(),
+      })
+      .overrideProvider(CurrencyManager)
+      .useValue({
+        assertCurrencyCompatibleWithCountry: jest.fn(),
+      })
+      .overrideProvider(PaypalCustomerManager)
+      .useValue({
+        createPaypalCustomer: jest.fn(),
+      })
+      .compile();
+
+    paypalService = moduleRef.get<PayPalService>(PayPalService);
+    paypalManager = moduleRef.get<PayPalManager>(PayPalManager);
+    currencyManager = moduleRef.get<CurrencyManager>(CurrencyManager);
+    paypalCustomerManager = moduleRef.get<PaypalCustomerManager>(
+      PaypalCustomerManager
+    );
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe('createBillingAgreement', () => {
+    it('creates a billing agreement', async () => {
+      // params
+      const uid = faker.string.uuid();
+      const token = faker.string.uuid();
+      const currency = faker.finance.currencyCode();
+
+      // faker responses
+      const billingAgreement: BillingAgreement = {
+        city: faker.location.city(),
+        countryCode: faker.location.countryCode(),
+        firstName: faker.person.firstName(),
+        lastName: faker.person.lastName(),
+        state: faker.location.state(),
+        status: BillingAgreementStatus.Active,
+        street: faker.location.streetAddress(),
+        street2: faker.location.streetAddress(),
+        zip: faker.location.zipCode(),
+      };
+
+      const paypalCustomer = {
+        uid: faker.string.hexadecimal({
+          length: 32,
+          prefix: '',
+          casing: 'lower',
+        }),
+        billingAgreementId: faker.string.hexadecimal({
+          length: 10,
+          prefix: '',
+        }),
+        status: 'active',
+        createdAt: faker.date.recent().getTime(),
+        endedAt: null,
+      };
```

Comment:
These types of things should come from factories rather than in-line within tests

Reply:
Updated!